[JIT] Introduce NamedType

jamesr66a opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #21839 [JIT] NamedTuple serialization
* #21836 [JIT] Refactor TupleType to take a FunctionSchema
* **#21835 [JIT] Introduce NamedType**
* #21813 [JIT] Some small fixes for NamedTuple

Differential Revision: [D15860004](https://our.internmc.facebook.com/intern/diff/D15860004)

Commit:
[JIT] Introduce SerializableType

Commit:
Update on "[JIT] Introduce SerializableType"

[JIT] Introduce SerializableType

gh-metadata: pytorch pytorch 21835 gh/jamesr66a/3/head

Commit:
Update on "[JIT] Introduce SerializableType"

[JIT] Introduce SerializableType

gh-metadata: pytorch pytorch 21835 gh/jamesr66a/3/head

Commit:
Update on "[JIT] Introduce SerializableType"

[JIT] Introduce SerializableType

gh-metadata: pytorch pytorch 21835 gh/jamesr66a/3/head

Commit:
Update on "[JIT] Introduce SerializableType"

[JIT] Introduce SerializableType

gh-metadata: pytorch pytorch 21835 gh/jamesr66a/3/head

suo commented on aten/src/ATen/core/jit_type.h:
```diff
@@ -1384,9 +1385,42 @@ struct ClassType;
+struct SerializableType;
+using SerializableTypePtr = std::shared_ptr<SerializableType>;
+
+struct CAFFE2_API SerializableType : public Type {
```
This should not inherit from Type. We don't want to muck up the type system with implementation details like this. Maybe make it a mixin

zdevito replied:
"NamedType" might be more accurate. Afterall, all of our types are serializable. I assume this will be come the shared parent for Module, Interface, and NamedTuple types?

Commit:
Update on "[JIT] Introduce SerializableType"

[JIT] Introduce SerializableType

gh-metadata: pytorch pytorch 21835 gh/jamesr66a/3/head

Commit:
Update on "[JIT] Introduce SerializableType"

[JIT] Introduce SerializableType

gh-metadata: pytorch pytorch 21835 gh/jamesr66a/3/head

Commit:
Update on "[JIT] Introduce SerializableType"

[JIT] Introduce SerializableType

gh-metadata: pytorch pytorch 21835 gh/jamesr66a/3/head

Commit:
Update on "[JIT] Introduce SerializableType"

[JIT] Introduce SerializableType

gh-metadata: pytorch pytorch 21835 gh/jamesr66a/3/head

zdevito commented on aten/src/ATen/core/jit_type.h:
```diff
@@ -1384,9 +1385,42 @@ struct ClassType;
+    return name_.name();
+  }
+
+  virtual std::vector<Function*> methods() const = 0;
```
What is the intention here? I don't image NamedTuple or Interfaces will have any methods.

jamesr66a replied:
Switched to using a `switch`

zdevito commented on aten/src/ATen/core/type.cpp:
```diff
@@ -468,23 +468,6 @@ bool Type::isSubtypeOf(const TypePtr rhs) const {
   return *this == *rhs;
 }
 
-ClassTypePtr ClassType::create(
```
I don't see the need for this change it is going to conflict with https://github.com/pytorch/pytorch/pull/21814. Can it be avoided?

jamesr66a replied:
These need to be moved out of c10 because otherwise they cause undefined reference to vtable of ClassType on android 🙄

Commit:
Update on "[JIT] Introduce SerializableType"

[JIT] Introduce SerializableType

gh-metadata: pytorch pytorch 21835 gh/jamesr66a/3/head

zdevito reviewed (approved):
NGNT

Commit:
Update on "[JIT] Introduce NamedType"

[JIT] Introduce SerializableType

gh-metadata: pytorch pytorch 21835 gh/jamesr66a/3/head
